### DISCUSSION TITLE
Exception handling in <T>LAPACK

### DISCUSSION THREAD

**Discussion by weslleyspereira** [tlapack/tlapack#48] [Ideas]
Currently, we also have the macros `blas_error_if` and `lapack_error_if` that can be disabled by `BLAS_ERROR_NDEBUG` and `NDEBUG`. The flags `NDEBUG` and `BLAS_ERROR_NDEBUG` represent 2 levels of exception handling at compile time. Recently, @thijssteel introduced some assertions on the array and block accesses in the legacy structures 8ccdc1a42de0f1f91087d53fe62d94bb9da670af. `assert` is a macro that is disabled when `NDEBUG` is defined. The release CMake build defines `NDEBUG`, so we would expect the asserts to be used only on no Release (Debug) modes.

I want to introduce some discussion topics so we can make progress on this thread:

1. @langou suggested we use different levels of checks in the code.
   For instance: inf-nan check, size check, array access check, parameter check. This could be implemented using different macros which would give us the control at compile time. What about control at runtime? One idea is to use the `struct opts_t` at the end of each routine to pass boolean attributes like `infNan_check`, `size_check`, etc. If the structure does not contain, say, `size_check` then the default behavior is the compile time behavior. If `struct opts_t` has `size_check` then there will be a check on the value of this boolean. I can try to work on this feature.

2. What about the behavior of the exception handler?
   We can either (1) return with a message / integer; or (2) throw an exception. Should we add flags to support both? I think there is no need to decide it at runtime, so the compiler flags will work.

3. Should we replace `blas::internal::error_if()` by `assert()` on the definition of `blas_error_if`?
   The real question is: Do we want to have an exception class in \<T\>LAPACK or do we just care about the message we display? If the latter holds, `assert` can do the job.

4. Should we have only one `tlapack_error_if` instead of `blas_error_if` and `lapack_error_if`? I think so, especially if we create the different levels of checks.

**Comment by thijssteel** [comment]
My two cents:

I think it's also worth distinguishing programming errors and true 'runtime' errors. The only way a size error is going to occur is if the programmer (either us or the user) makes a mistake. Those can definitely be assertions and thus be turned off in production builds. The inf checks, and certainly something like a convergence failure in the iterative algorithms should really be thrown like real exceptions and handled correctly in the calling routine.

**Reply by weslleyspereira** [reply]
About the size errors: what about matrices loaded at runtime? A program will need to check at runtime that the sizes match before using, say, gemm.

**Reply by thijssteel** [reply]
I still view that as a programming error, if you load a matrix from some kind of storage or user input, you should check it yourself before calling a LAPACK routine. The assertions we put in place can help with debugging.

**Comment by weslleyspereira** [comment]
Another type of check that I could think of is:

- On potrs, A must not share memory with B. However, nothing prevents the user of calling `potrs(uplo,A,A)`. So, is there an effective way to check if two arrays share the same memory before performing one operation?
- I could thought of using `lapack_error_if( std::addressof(A) == std::addressof(B) )` but it does not cover all cases. For instance, `std::addressof(A) != std::addressof(B)` if B is a slice of A.

__[edit:]__
- I am not sure it is worth doing these checks now. Also, LAPACK does not do such tests.
- I think it is fair to require that: _Whenever A and B are output arguments, A is assumed to not share memory with B_

**Reply by langou** [reply]
Yes, same for GEMM. In Matlab, you can type A = A*B and that works. If you call GEMM( A, A, B), the behavior is undefined.

**Reply by thijssteel** [reply]
It's technically possible to check for aliasing I guess, but that would need to be specialized for the matrix types (because matrices are not contiguous in memory)

**Reply by weslleyspereira** [reply]
The C++ standard uses the documentation to require non-overlapping data. E.g., function `std::copy`: https://en.cppreference.com/w/cpp/algorithm/copy:
- "The behavior is undefined if the source and the destination ranges overlap."
- "When copying overlapping ranges, std::copy is appropriate when copying to the left (beginning of the destination range is outside the source range) while std::copy_backward is appropriate when copying to the right (end of the destination range is outside the source range)."

**Reply by langou** [reply]
Yes, this is one reason why having the notion of "in" / "inout" / "out" is useful. If two pieces of data are "in", then it is OK to overlap them. For example, using GEMM to do B = A^T A would work. (Although using SYRK is probably a better idea!)

**Comment by langou** [comment]
I do not think we should worry about this for now. 

Such features are also useful for task parallelism. We label array arguments of tasks are in, or inout, or out, then a runtime can schedule them. To do this, it is better to know the extent of the arrays, not just the pointer.

**Comment by weslleyspereira** [comment]
After a discussion with @thijssteel about [9b98605](https://github.com/tlapack/tlapack/pull/55/commits/9b98605308569e483ef705fac9b31772b8ad6f3a), I have some questions:

1. what would be the use case of splitting the error checks for: input params, array sizes and access policies?

2. do we really need enable/disable flags for all kinds of error checks? Or should we define some of them using only compiler flags, e.g., size and param checks like in LAPACK currently?

3. what about having the control of enable/disable the checks by groups of routines? For instance: (1) lv1 BLAS, (2) lv2 BLAS, (3) lv3 BLAS, (4) LAPACK.

And @thijssteel give one idea that I would like to try: split `opts` from `exception_handler_opts` (or a shorter name) in the interface.

**Reply by thijssteel** [reply]
Splitting opts is to make it easier to propagate the flags. Otherwise you would need to pass the entire opts object (with possibly conflicting parameters) or recreate the opts object for each routine you want to call.

**Reply by weslleyspereira** [reply]
Thanks for explaining!
